rust/core-lib/src/width_cache.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Cache and utilities for doing width measurement.
+
+use std::borrow::Cow;
+use std::collections::{BTreeMap, HashMap};
+
+use xi_rpc;
+
+use tabs::{DocumentCtx, WidthReq};
+
+/// A token which can be used to retrieve an actual width value when the
+/// batch request is submitted.
+/// 
+/// Internally, it is implemented as an index into the `widths` array.
+pub type Token = usize;
+
+pub struct WidthCache {
+    // maps cache key to index within widths
+    m: HashMap<WidthCacheKey<'static>, Token>,
+    widths: Vec<f64>,
+}
+
+#[derive(Eq, PartialEq, Hash)]
+struct WidthCacheKey<'a> {
+    id: usize,  // style id
+    s: Cow<'a, str>,
+}
+
+/// A batched request, so that a number of strings can be measured in a
+/// a single RPC.
+pub struct WidthBatchReq<'a> {
+    cache: &'a mut WidthCache,
+    pending_tok: usize,
+    req: Vec<WidthReq>,
+    req_toks: Vec<Vec<Token>>,
+    // maps style id to index into req/req_toks
+    req_ids: BTreeMap<usize, usize>,
+}
+
+impl WidthCache {
+    pub fn new() -> WidthCache {
+        WidthCache {
+            m: HashMap::new(),
+            widths: Vec::new(),
+        }
+    }
+
+    /// Resolve a previously obtained token into a width value.
+    pub fn resolve(&self, tok: Token) -> f64 {
+        self.widths[tok]
+    }
+
+    /// Create a new batch of requests.
+    pub fn batch_req(self: &mut WidthCache) -> WidthBatchReq {
+        let pending_tok = self.widths.len();
+        WidthBatchReq {
+            cache: self,
+            pending_tok,
+            req: Vec::new(),
+            req_toks: Vec::new(),
+            req_ids: BTreeMap::new(),
+        }
+    }
+}
+
+impl<'a> WidthBatchReq<'a> {
+
+    /// Request measurement of one string/style pair within the batch.
+    // Note: we could immediately return a width here on cache hit, or
```

Comment:
It was originally on purpose, as a "note to self", but I ended up deleting it, after having written more code I don't think it would make things better either for clarity or efficiency.